### PR TITLE
B/command

### PR DESCRIPTION
**Pull request by uglrl**
# Pull Request for Implementation

## Backend
Make sure, you have done de following before merging this pull request:
- [x] Your code is **documented** using doc comments.
- [x] You added appropriate **logging** in your code.
- [x] You added **unit tests** to all your implemented functions.
- [x] You ran `cargo format` to format your code.
- [x] You ran `cargo clippy` to check for improvements.
- [ ] You added the **class descriptions** of the classes you implemented in the **latex document** on overleaf.
      Make sure to mark any changes you made to the design documents specification.

## TODO
- [x] Test for CommandHandler and Authenticator
- [x] Better enums for Authenticator

### CODE REVIEW

File: backend/src/layer/logic/api_command/command_handler.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+#[async_trait]
+impl<D, A, I> Command for CommandHandler<D, A, I>
+where
+    D: CommandDataAccess + Sync + Send,
+    A: AdminNotification + Sync + Send,
+    I: ImageHoster + Sync + Send,
+{
+    /// Command to report an image. It als gets checked whether the image shall get hidden.
```
Comment (worldofjoni) [inline backend/src/layer/logic/api_command/command_handler.rs:80]: als**o**

File: backend/src/layer/logic/api_command/command_handler.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    async fn report_image(
+        &self,
+        image_id: Uuid,
+        reason: ReportReason,
+        auth_info: AuthInfo,
+    ) -> Result<()> {
+        if let Some(auth_info) = auth_info {
```
Comment (worldofjoni) [inline backend/src/layer/logic/api_command/command_handler.rs:87]: directly return when auth_info is none

File: backend/src/layer/logic/api_command/command_handler.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+            let image_command_type = ImageCommandType::ReportImage(reason);
+            self.auth
+                .authn_image_command(&auth_info, image_id, image_command_type)?;
+            let info = self.command_data.get_image_info(image_id).await?;
+            if !info.approved {
+                self.command_data
+                    .add_report(image_id, auth_info.client_id, reason)
+                    .await?;
+                let image_got_hidden = Self::will_be_hidden(info.upload_date, info.report_count);
```
Comment (worldofjoni) [inline backend/src/layer/logic/api_command/command_handler.rs:96]: should that not rather be called `should_hide` as the image did not already get hidden?

File: backend/src/layer/logic/api_command/command_handler.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                if image_got_hidden {
+                    self.command_data.hide_image(image_id).await?;
+                }
+                let report_info = ImageReportInfo {
+                    reason,
+                    image_id,
+                    image_got_hidden,
+                    image_link: info.image_url,
+                    report_count: info.report_count,
```
Comment (worldofjoni) [inline backend/src/layer/logic/api_command/command_handler.rs:105]: Mind thath the current report is not part of the info gotten from the database.
Maybe there is a better way (better place) to add this report, as this also has to be considerd inside `Self::will_get_hidden`. Maybe just increment the counter on the info.

Comment (worldofjoni) [reply backend/src/layer/logic/api_command/command_handler.rs:105]: as the additional report also may influence the image rank, maybe we should re-request the image info from the database modifications?
or we should make clear that these stats are the state before the report?

File: backend/src/layer/logic/api_command/command_handler.rs (hunk 1 above)
Comment (worldofjoni) [reply backend/src/layer/logic/api_command/command_handler.rs:80]: these documentation comments should get removed as they are already specified in the interface.

File: backend/src/layer/logic/api_command/command_handler.rs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+    /// Command to link an image to a meal.
+    async fn add_image(&self, meal_id: Uuid, image_url: String, auth_info: AuthInfo) -> Result<()> {
+        if let Some(auth_info) = auth_info {
+            self.auth
+                .authn_add_image_command(&auth_info, meal_id, &image_url)?;
+            let image_meta_data = self.image_hoster.validate_url(image_url).await?;
+            let licence_ok = self.image_hoster.check_licence(image_meta_data.id).await?;
+            if licence_ok {
+                //self.command_data.link_image(meal_id, auth_info.client_id, image_meta_data.id, image_meta_data.image_url).await;
```
Comment (worldofjoni) [inline backend/src/layer/logic/api_command/command_handler.rs:189]: what happened here?

File: backend/src/layer/logic/api_command/command_handler.rs [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+                return Ok(());
+            }
+            todo!()
```
Comment (worldofjoni) [inline backend/src/layer/logic/api_command/command_handler.rs:192]: todo?

Comment (worldofjoni) [reply backend/src/layer/logic/api_command/command_handler.rs:192]: you may create a new error variant for this case.

File: backend/src/layer/trigger/graphql/bin.rs [hunk 7]
```diff
@@ -16,7 +22,17 @@ async fn main() {
...
+    let mut server = server::GraphQLServer::new(
+        info,
+        mock::RequestDatabaseMock,
+        CommandHandler::new(
+            CommandDatabaseMock,
+            CommandAdminNotificationMock,
+            CommandImageHosterMock,
+        )
+        .await
+        .expect("could not craete command mock"),
```
Comment (worldofjoni) [inline backend/src/layer/trigger/graphql/bin.rs:34]: create

File: backend/src/layer/logic/api_command/auth/authenticator.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+use sha2::{Digest, Sha512};
+
+use crate::{
+    interface::api_command::{CommandError, InnerAuthInfo, Result},
+    util::Uuid,
+};
+
+use super::image_command_type::ImageCommandType;
+
+const MEAL_RATING_COMMAND_NAME: &str = "setRating"; // TODO
```
[the remark this answers is not in the document]
Comment (uglrl) [reply backend/src/layer/logic/api_command/auth/authenticator.rs:12]: ?

File: backend/src/interface/api_command.rs [hunk 9]
```diff
@@ -43,7 +44,7 @@ pub trait Command {
 pub type AuthInfo = Option<InnerAuthInfo>;
 
 /// Structure containing all information necessary for authenticating a client.
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Comment (Whatsuup) [inline backend/src/interface/api_command.rs:47]: Clone -> Copy, Clone

File: backend/src/layer/logic/api_command/auth/authenticator.rs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            reason.to_string()
+        } else {
+            String::new()
+        };
+
+        let hash = self.calculate_hash(
+            auth_info,
+            &image_command_type.to_string(),
+            &[&image_id, &report_reason],
+        )?;
```
Comment (Whatsuup) [inline backend/src/layer/logic/api_command/auth/authenticator.rs:43]: Calculate hash in external function. `gen_hash_image_cmd(...) -> shall`. Thinks like that could be repetitive and usually be determinded isolated. For changes in generation, it's also better to change the functions as each part in the auth functions.

File: backend/src/layer/logic/api_command/command_handler.rs [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+};
+
+const REPORT_FACTOR: f64 = 1.0 / 35.0;
+
+pub struct CommandHandler<D, A, I>
+where
+    D: CommandDataAccess + Sync + Send,
+    A: AdminNotification + Sync + Send,
+    I: ImageHoster + Sync + Send,
+{
```
Comment (Whatsuup) [inline backend/src/layer/logic/api_command/command_handler.rs:25]: Change generic letters to words (e.g. D -> DataAccess)

**Review by Whatsuup — CHANGES_REQUESTED** [review]
No tests exist so far. Is this intended? The authenticator provides functions that could be interesting...

File: backend/src/interface/api_command.rs (hunk 9 above)
Comment (worldofjoni) [reply backend/src/interface/api_command.rs:47]: cannot / should not be copy as a String is not copy

**Comment by worldofjoni** [comment]
> No tests exist so far. Is this intended? The authenticator provides functions that could be interesting...
Test will be added

File: backend/src/layer/logic/api_command/auth/authenticator.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (worldofjoni) [reply backend/src/layer/logic/api_command/auth/authenticator.rs]: i will work on that

**Comment by worldofjoni** [comment]
@uglrl tests for command handler are still missing, then @Whatsuup can do a final review so that we can merge.

File: backend/src/layer/logic/api_command/auth/authenticator.rs [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+            "YWpzZGg4MnozNzhkMnppZGFzYXNkMiBzYWZzYSBzPGE5MDk4".into()
+        ]);
+
+        let info = InnerAuthInfo {
+            api_ident: "YWpzZGg4Mn".into(),
+            hash: "lb4TH+zjHTl0Z9zijEZ7KtOFIBFHvY70rmZtX+Xk/fa++fGJtAS10EjFOqAgx/0scDJDbhpdn9WS5Yy5zCYeoQ==".into(),
+            client_id: Uuid::try_from("4c57fc70-4839-4398-be08-d151c0dbb246").unwrap(),
+        };
+
+        let command = CommandType::RemoveDownvote {
```
Comment (uglrl) [inline backend/src/layer/logic/api_command/auth/authenticator.rs:178]: This should be RemoveUpvote and not RemoveDownvote

**Review by Whatsuup — APPROVED** [review]
As always, don't forget to document your changes in overleaf!

**Pull request merged**